src/preferences/dialog/dlgprefeffects.cpp

```diff
@@ -12,53 +12,79 @@ DlgPrefEffects::DlgPrefEffects(QWidget* pParent,
+            SLOT(availableEffectsListItemSelected(const QModelIndex&)));
+
+    // Highlight first row
+    availableEffectsList->selectRow(0);
+
+  #if QT_VERSION < QT_VERSION_CHECK(5, 0, 0)
+    availableEffectsList->horizontalHeader()->setResizeMode(0, QHeaderView::ResizeToContents);
+    availableEffectsList->setColumnWidth(1, 200);
+    availableEffectsList->horizontalHeader()->setResizeMode(2, QHeaderView::ResizeToContents);
+  #else
+    availableEffectsList->horizontalHeader()->setSectionResizeMode(0, QHeaderView::ResizeToContents);
+    availableEffectsList->setColumnWidth(1, 200);
+    availableEffectsList->horizontalHeader()->setSectionResizeMode(2, QHeaderView::ResizeToContents);
+  #endif // QT_VERSION < QT_VERSION_CHECK(5, 0, 0)
+}
+
+DlgPrefEffects::~DlgPrefEffects() {
+    delete m_pAvailableEffectsModel;
 }
 
 void DlgPrefEffects::slotUpdate() {
     clear();
-    const QList<EffectManifestPointer> availableEffectManifests =
-            m_pEffectsManager->getAvailableEffectManifests();
-
-    for (const auto& pManifest : availableEffectManifests) {
-        QListWidgetItem* pItem = new QListWidgetItem();
-        pItem->setData(Qt::UserRole, pManifest->id());
-        pItem->setText(pManifest->displayName());
-        availableEffectsList->addItem(pItem);
-    }
+    m_pAvailableEffectsModel->resetFromEffectManager(m_pEffectsManager);
 
-    if (!availableEffectManifests.isEmpty()) {
-        availableEffectsList->setCurrentRow(0);
+    if (!m_pAvailableEffectsModel->isEmpty()) {
+        availableEffectsList->selectRow(0);
     }
 }
 
 void DlgPrefEffects::slotApply() {
-    // Nothing to apply.
+    for (EffectProfilePtr profile : m_pAvailableEffectsModel->profiles()) {
+        EffectManifestPointer pManifest = profile->pManifest;
+        m_pEffectsManager->setEffectVisibility(pManifest, profile->bIsVisible);
+        m_pConfig->set(ConfigKey("[Visible Effects]", pManifest->id()), ConfigValue(profile->bIsVisible));
+    }
 }
 
 void DlgPrefEffects::slotResetToDefaults() {
-    // Nothing to reset.
+    slotUpdate();
 }
 
 void DlgPrefEffects::clear() {
-    availableEffectsList->clear();
     effectName->clear();
     effectAuthor->clear();
     effectDescription->clear();
     effectVersion->clear();
     effectType->clear();
 }
 
-void DlgPrefEffects::slotEffectSelected(QListWidgetItem* pCurrent,
-                                        QListWidgetItem* pPrevious) {
-    Q_UNUSED(pPrevious);
-    if (pCurrent == NULL) {
+void DlgPrefEffects::availableEffectsListItemSelected(const QModelIndex& selected) {
+    QString effectId = m_pAvailableEffectsModel->data(selected, Qt::UserRole).toString();
+
+    if (effectId == QVariant().toString())
         return;
-    }
-    QString effectId = pCurrent->data(Qt::UserRole).toString();
+
     EffectManifestPointer pManifest;
     EffectsBackend* pBackend;
     m_pEffectsManager->getEffectManifestAndBackend(effectId, &pManifest, &pBackend);
```

Comment:
GitHub won't let me comment directly on the lines below because you haven't modified them yet. Anyway, use `effectType->setText(EffectManifest::backendTypeToTranslatedString(pManifest->backendType())` below instead of `pBackend->getName()`.

I noticed an issue with the `EffectsManager::getEffectManifestAndBackend` and `EffectsManager::getEffectManifest` functions. If two backends each have an effect that happens to have the same ID string, there will be conflict between them and only the effect in the backend that happened to be registered first will be available to the rest of Mixxx. Currently this is a hypothetical issue because LV2 effects are identified by a URL which wouldn't conflict with the ID strings for built-in effects (assuming a standards-compliant LV2 plugin, which we cannot assume because we do not control the LV2 plugins), so fixing this is not a big priority and does not need to be done in this PR. For now, removing DlgPrefEffects' dependency on `EffectsManager::getEffectManifestAndBackend` is enough. I think we should keep this issue in mind going forward with the refactoring though. Effects should be identified by both their EffectType and a QString, not a QString alone.

Reply:
I've removed the dependency on `EffectsManager::getEffectManifestAndBackend` in the last commit.

I noticed the same a while ago. I think we can create a hash function using the EffectType and the Effect Id, and use it rather than just the Effect Id.

Reply:
We may consider adding an equality operator to the EffectManifest class that compares both the ID string and the EffectBackendType.